test_plans/reduced_feature_set.asciidoc

```diff
@@ -0,0 +1,45 @@
+
+A CTS option is introduced, `SYCL_CTS_ENABLE_FEATURE_SET_FULL`, which specifies that full features should be tested in addition to the features of the reduced feature set. The tests for the full features should be guarded if `#ifdef SYCL_CTS_ENABLE_FEATURE_SET_FULL`.
+
+In all existing tests, the full features are removed and replaced with a workaround. Since the CTS was not written with the feature sets in mind, a "best effort" is made to remove the full features in the CTS. New additions should not use full features unless guarded with `SYCL_CTS_ENABLE_FEATURE_SET_FULL`.
+
+An exception to this are the tests that explicitly test full features. For example, when launching a kernel to test `sycl::id` 's interface, the kernel should be launched with a kernel name instead of using the full feature to not specify a name. However, the test that verifies that the full feature that allows for kernels to be launched without a kernel name works correctly, should still exist and be guarded with `SYCL_CTS_ENABLE_FEATURE_SET_FULL`.
+
+Currently, three full features are defined. For each of these features, the required behavior as a result from removing full features is stated, in addition to the tests which are the exceptions (or the proposal for a simple test if none exist).
+
+== Un-named SYCL kernel functions
+=== Required behavior
+Every kernel launch has a kernel name, e.g. `parallel_for<kernel_name>`.
+
+=== Exceptions
+The existing tests for `sycl::handler` in `handler/handler_invoke_api.cpp` are modified such that they are also run without a kernel name, if `SYCL_CTS_ENABLE_FEATURE_SET_FULL` is defined.
+
+== Generic address space mode
+=== Required behavior
+Section 5.9.4. of the SYCL 2020 specification defines several rules for the deduction of the address space of pointer types and reference types. It is illegal when following these rules results in a pointer value being assigned to a pointer variable that addresses a different addressing space. Tests that depend on such behavior should be rewritten.
+
+=== Exceptions
+Existing tests rely on the generic address space being available. These tests are guarded with `SYCL_CTS_ENABLE_FEATURE_SET_FULL`:
+
+- `multi_ptr_convert_assignment_ops_(core|fp16|fp64).cpp`
+- `multi_ptr_explicit_conversions_(core|fp16|fp64).cpp`
+- `multi_ptr_implicit_conversions_(core|fp16|fp64).cpp`
+
+== ODR-usage of static declarations
+=== Required behavior
+No variables with static storage duration may be ODR-used inside a device function.
+
+=== Exceptions
+A simple test is constructed that ODR-uses a variable with static storage duration and produces some side effect. The test is guarded with `SYCL_CTS_ENABLE_FEATURE_SET_FULL`.
+
+- Declare `static const int value = <value>;`
+- Launch a kernel that takes the address of `value` (which ODR-uses the variable) and stores the result.
+- Check that the result does not evaluate to `nullptr`.
```

Comment:
If we return the address back to the host (e.g. via a buffer), then the host can check that the address is the same as `&value`.

Shall we also do this test for a `static constexpr` variable?

Reply:
> If we return the address back to the host (e.g. via a buffer), then the host can check that the address is the same as `&value`.

That was my assumption as well, but I tested it with DPCPP and found that these are not equal. I think that makes sense, since the address spaces of the device and host are not (necessarily) the same. I too find it unfortunate that the value cannot be checked. I could not come up with a test that 1. uses a variable with static storage duration, 2. the kernel ODR-uses the variable, and 3. the result can be checked.

> Shall we also do this test for a `static constexpr` variable?

I have added it to the test plan.

Reply:
Maybe this is a better test about ODR-using a static variable that is `static const` or `static constexpr`:

* Define a kernel that takes the address of `value` and stores the result.
* From the same kernel, read the stored address, dereference the pointer, and make sure the value returned is `<value>`.  Structure the code such that the compiler cannot determine statically that the stored address is really the address of `value`.

I think our intent is that it is legal to take the address of the variable from device code and using it within device code, but we do not make any guarantee about using the address on the host.

Reply:
Okay, that seems to work. I have updated the test plan.